### PR TITLE
feat: Add babylon-related metrics

### PR DESCRIPTION
**Pull request by gitferry**
This PR adds Babylon-related metrics for monitoring purposes. The metrics and alarming rules are described as follows. cc @philmln.

* reporter
    * "vigilante_reporter_reported_headers", "The total number of reported BTC headers to Babylon"
    * "vigilante_reporter_reported_checkpoints", "The total number of reported BTC checkpoints to Babylon"
    * "vigilante_reporter_failed_headers", "The total number of failed BTC headers to Babylon", should alarm when this metric > 0
    * "vigilante_reporter_failed_checkpoints", "The total number of failed BTC checkpoints to Babylon", should alarm when this metric > 0
    * "vigilante_reporter_since_last_header_seconds", "Seconds since the last successful reported BTC checkpoint to Babylon", should alarm when this metric > X
    * "vigilante_reporter_since_last_submission_seconds", "Seconds since the last successful reported BTC checkpoint to Babylon", should alarm when this metric > X
* submitter
    * "vigilante_submitter_submitted_checkpoints", "The total number of submitted checkpoints to BTC"
    * "vigilante_submitter_failed_checkpoints", "The total number of failed checkpoints to BTC", should alarm when this metric > 0
    * "vigilante_submitter_since_last_checkpoint_seconds", "Seconds since the last successfully submitted checkpoint", should alarm when this metric > X
* monitor
    * "vigilante_monitor_valid_epochs", "The total number of verified epochs"
    * "vigilante_monitor_invalid_epochs", "The total number of invalid epochs", should alarm when this metric > 0
    * "vigilante_monitor_valid_btc_headers", "The total number of valid BTC headers"
    * "vigilante_monitor_invalid_epochs", "The total number of invalid BTC headers", should alarm when this metric > 0
    * "vigilante_monitor_liveness_attacks", "The total number of detected liveness attacks", should alarm when this metric > 0

### CODE REVIEW

**Review by vitsalis — APPROVED** [review]
Nice! Those are all very useful metrics.

Note that we should not alarm when the number of failed checkpoints submitted to Babylon surpasses `0`. One can insert a bad checkpoint on purpose.

File: metrics/monitor.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+func NewMonitorMetrics() *MonitorMetrics {
+	registry := prometheus.NewRegistry()
+	registerer := promauto.With(registry)
+
+	metrics := &MonitorMetrics{
+		Registry: registry,
+		ValidEpochsCounter: registerer.NewCounter(prometheus.CounterOpts{
+			Name: "vigilante_monitor_valid_epochs",
+			Help: "The total number of verified epochs",
```
Comment (SebastianElvis) [inline metrics/monitor.go:25]:
```suggestion
			Help: "The total number of valid epochs",
```

File: metrics/monitor.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+		}),
+		ValidBTCHeadersCounter: registerer.NewCounter(prometheus.CounterOpts{
+			Name: "vigilante_monitor_valid_btc_headers",
+			Help: "The total number of valid BTC headers",
+		}),
+		InvalidBTCHeadersCounter: registerer.NewCounter(prometheus.CounterOpts{
+			Name: "vigilante_monitor_invalid_epochs",
+			Help: "The total number of invalid BTC headers",
+		}),
+		LivenessAttacksCounter: registerer.NewCounter(prometheus.CounterOpts{
```
Comment (SebastianElvis) [inline metrics/monitor.go:39]: Since the monitor will just panick upon a liveness attack, do we need this metrics?

File: metrics/monitor.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+type MonitorMetrics struct {
+	Registry                 *prometheus.Registry
+	ValidEpochsCounter       prometheus.Counter
+	InvalidEpochsCounter     prometheus.Counter
+	ValidBTCHeadersCounter   prometheus.Counter
+	InvalidBTCHeadersCounter prometheus.Counter
+	LivenessAttacksCounter   prometheus.Counter
```
Comment (SebastianElvis) [inline metrics/monitor.go:14]: Should these be pointers or objects?

File: metrics/reporter.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+)
+
+type ReporterMetrics struct {
+	Registry                        *prometheus.Registry
+	SuccessfulHeadersCounter        prometheus.Counter
+	SuccessfulCheckpointsCounter    prometheus.Counter
+	FailedHeadersCounter            prometheus.Counter
+	FailedCheckpointsCounter        prometheus.Counter
+	SecondsSinceLastHeaderGauge     prometheus.Gauge
+	SecondsSinceLastCheckpointGauge prometheus.Gauge
```
Comment (SebastianElvis) [inline metrics/reporter.go:17]: same question here

File: metrics/submitter.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+type SubmitterMetrics struct {
+	Registry                        *prometheus.Registry
+	SuccessfulCheckpointsCounter    prometheus.Counter
+	FailedCheckpointsCounter        prometheus.Counter
+	SecondsSinceLastCheckpointGauge prometheus.Gauge
```
Comment (SebastianElvis) [inline metrics/submitter.go:14]: same ques here

File: metrics/reporter.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+			Help: "Seconds since the last successful reported BTC checkpoint to Babylon",
+		}),
+	}
+	return metrics
+}
+
+func (sm *ReporterMetrics) RecordMetrics() {
+	go func() {
+		for {
+			time.Sleep(1 * time.Second)
```
Comment (SebastianElvis) [inline metrics/reporter.go:57]: We can use [ticker](https://gobyexample.com/tickers) here

File: metrics/reporter.go [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+			// will be reset when a header/checkpoint is successfully sent
+			sm.SecondsSinceLastHeaderGauge.Inc()
+			sm.SecondsSinceLastCheckpointGauge.Inc()
```
Comment (SebastianElvis) [inline metrics/reporter.go:60]: Given that the two gauges will never decrease, maybe they can be counters?

**Review by SebastianElvis — APPROVED** [review]
This is a great improvement! Just some minor comments

File: metrics/monitor.go (hunk 3 above)
Comment (gitferry) [reply metrics/monitor.go:14]: These are interfaces, so it depends on whether the actual values are pointers or objects.

File: metrics/monitor.go (hunk 2 above)
Comment (gitferry) [reply metrics/monitor.go:39]: We don't want to panic the monitor or the vigilante but send an alert instead and let the operator decide how to deal with this.

File: metrics/reporter.go (hunk 7 above)
Comment (gitferry) [reply metrics/reporter.go:60]: They will be reset when the last header/checkpoint is successfully sent

File: metrics/reporter.go (hunk 6 above)
Comment (gitferry) [reply metrics/reporter.go:57]: I'm not sure how to use ticker here. Btw, why can't we use sleep here?

Comment (SebastianElvis) [reply metrics/reporter.go:57]: It basically achieves the same functionality as sleeping here, but a more formal way.
Something like below
```go
	ticker := time.NewTicker(1 * time.Second)
	for range ticker.C {
    	// do our stuff
   	}
```

You can find multiple usages of this in vigilante repo.

Comment (gitferry) [reply metrics/reporter.go:57]: I see. It's good to know. Thanks a lot!

**Pull request merged**